src/openzeppelin/tests/test_dual721.cairo

```diff
@@ -0,0 +1,525 @@
+use traits::Into;
+use array::ArrayTrait;
+use starknet::ContractAddress;
+use starknet::contract_address_const;
+use starknet::testing::set_caller_address;
+use starknet::testing::set_contract_address;
+use openzeppelin::token::erc721::interface::IERC721Dispatcher;
+use openzeppelin::token::erc721::interface::IERC721CamelDispatcher;
+use openzeppelin::token::erc721::interface::IERC721DispatcherTrait;
+use openzeppelin::token::erc721::interface::IERC721CamelDispatcherTrait;
+use openzeppelin::token::erc721::dual721::DualCaseERC721Trait;
+use openzeppelin::token::erc721::dual721::DualCaseERC721;
+use openzeppelin::tests::mocks::snake721_mock::SnakeERC721Mock;
+use openzeppelin::tests::mocks::camel721_mock::CamelERC721Mock;
+use openzeppelin::tests::mocks::erc721_receiver::ERC721Receiver;
+use openzeppelin::tests::mocks::erc721_receiver::SUCCESS;
+use openzeppelin::tests::mocks::erc721_receiver::FAILURE;
+use openzeppelin::tests::mocks::erc721_panic_mock::SnakeERC721PanicMock;
+use openzeppelin::tests::mocks::erc721_panic_mock::CamelERC721PanicMock;
+use openzeppelin::tests::mocks::non_implementing_mock::NonImplementingMock;
+use openzeppelin::tests::utils;
+
+///
+/// Constants
+///
+
+const NAME: felt252 = 111;
+const SYMBOL: felt252 = 222;
+const URI: felt252 = 333;
+
+fn TOKEN_ID() -> u256 {
+    7.into()
+}
+fn OWNER() -> ContractAddress {
+    contract_address_const::<10>()
+}
+fn RECIPIENT() -> ContractAddress {
+    contract_address_const::<20>()
+}
+fn SPENDER() -> ContractAddress {
+    contract_address_const::<30>()
+}
+fn OPERATOR() -> ContractAddress {
+    contract_address_const::<40>()
+}
+fn DATA(success: bool) -> Span<felt252> {
+    let mut data = ArrayTrait::new();
+    if success {
+        data.append(SUCCESS);
+    } else {
+        data.append(FAILURE);
+    }
+    data.span()
+}
+
+///
+/// Setup
+///
+
+fn setup_snake() -> (DualCaseERC721, IERC721Dispatcher) {
+    let mut calldata = ArrayTrait::new();
+    calldata.append(NAME);
+    calldata.append(SYMBOL);
+    calldata.append(TOKEN_ID().low.into());
+    calldata.append(TOKEN_ID().high.into());
+    calldata.append(URI);
+    set_caller_address(OWNER());
```

Comment:
I think is important to mention that this is a workaround for a discrepancy **(bug? could be intentional)** between the [cairo-lang-runner](https://github.com/starkware-libs/cairo/blob/eaf16c6ebfb5ddc329155a039d764d99dd507d9f/crates/cairo-lang-runner/src/casm_run/mod.rs#L797) and the actual [execution module for the sequencer (blockifier)](https://github.com/starkware-libs/blockifier/blob/a3a133af7eaa07197caa90766db11edacf8f46f6/crates/blockifier/src/execution/syscalls/mod.rs#L238).

In order to be closer to reality, we should be using `set_contract_address(OWNER())` here instead, because the `deploy_syscall` also (as `call_contract_syscall`) changes the context for the caller (in the sequencer), but in the compiler repo is not doing that.

I think we can merge like this, but we need to keep this in mind, and maybe submit a PR to the cairo-lang-runner crate for fixing the discrepancy (is fairly easy).

Reply:
This affects basically every call to the `deploy_syscall` in tests.

Reply:
For clarity, the bug is that the compiler crate for execution is not changing the caller_address in context when executing a deploy syscall, while the actual sequencer code is doing it, as expected. So we are setting caller address before calling `deploy_syscall`, for setting the caller, when we should be setting contract address instead for that.

Reply:
cc @andrew-fleming

Reply:
And I agree, we can merge like this